ipyvolume/embed.py

```diff
@@ -140,63 +57,118 @@ def save_js(dirname, makedirs=True):
-        while add_referring_widgets(state):
-            pass
-
-        values = template_options.copy()
-        values.update(dict(extra_script_head="", body_pre="", body_post=""))
-        values.update(kwargs)
-        widget_views = ""
-        for widget in widgets:
-            widget_views += widget_view_template.format(**dict(model_id=widget.model_id))
-        # Rely on ipywidget to get the default values
-        json_data = ipywidgets.widgets.Widget.get_manager_state(widgets=[])
-        # but plug in our own state
-        json_data['state'] = state
-        if external_json:
-            filename_base = os.path.splitext(filename)[0]
-            with open(filename_base+".json", "w") as fjson:
-                json.dump(json_data, fjson)
-            values.update(dict(title=title, widget_views=widget_views))
-        else:
-            values.update(dict(title=title,
-                      json_data=json.dumps(json_data, indent=indent),
-                           widget_views=widget_views))
-        html_code = template.format(**values)
-        f.write(html_code)
 
-    if copy_js:
-        save_js(dir_name_dst, makedirs=makedirs)
-        
-    
+    template_opts = dict(template_options)
+    template_opts['title'] = title
+
+    if all_states:
+        state = None
+    else:
+        state = wembed.dependency_state(widgets, drop_defaults=drop_defaults)
+
+    if not offline:
+        template_opts['snippet'] = wembed.embed_minimal_html(filepath, widgets, state=state,
+                                                             requirejs=True, drop_defaults=drop_defaults)
+    else:
+        if offline_req:
+            scripts_path = os.path.join(dir_name_dst, offline_folder)
+            if not os.path.exists(scripts_path):
+                os.makedirs(scripts_path)
+            # TODO embed-amd.js looks for ipyvolume.js in the local path of the html file, not it's own local path,
+            # so this can't be in the scripts path
+            save_ipyvolumejs(dir_name_dst)
+            save_requirejs(os.path.join(scripts_path, "require.min.js"))
+            save_embed_js(os.path.join(scripts_path, "embed-amd.js"))
+            save_font_awesome(os.path.join(scripts_path, "font-awesome"))
+
+        offline_folder = ''
+        if offline_folder:
+            offline_folder += '/'
+        snippet = wembed.embed_snippet(widgets, embed_url=offline_folder+"embed-amd.js",
+                                       requirejs=False, drop_defaults=drop_defaults, state=state)
+        offline_snippet = """
+        <link href="{offline_folder}font-awesome/css/font-awesome.min.css" rel="stylesheet">    
+        <script src="{offline_folder}require.min.js" crossorigin="anonymous"></script>
+        {snippet}
+        """.format(offline_folder=offline_folder, snippet=snippet)
+
+        template_opts['snippet'] = offline_snippet
+
+    fmt = DefaultFormatter()
+    html_code = fmt.format(template, **template_opts)
```

Comment:
I'd keep the {}.update() strategy. Less people would know what the defaultformatter does, would have to look it up etc, I'd say it goes agains Python zen :smile: